go/database/mpt/nodes_test.go

```diff
@@ -6554,6 +6554,162 @@ func markModifiedAsDirty(t *testing.T, ctxt *nodeContext, before, after NodeRefe
 	handle.Release()
 }
 
+func TestVisitPathToAccount_CanIterateNodesByCorrectAddresses(t *testing.T) {
+	address := common.Address{0x12, 0x34, 0x56, 0x78}
+
+	tests := map[string]struct {
+		trie NodeDesc // < the structure of the trie
+		path []string // < the path to follow to reach the test account
+	}{
+		"empty": {
+			trie: &Tag{"A", &Empty{}},
+			path: []string{"A"},
+		},
+		"wrong account": {
+			trie: &Tag{"A", &Account{}},
+			path: []string{"A"}, // < this should be empty
+		},
+		"correct account": {
+			trie: &Tag{"A", &Account{address: address}},
+			path: []string{"A"},
+		},
+		"branch without account": {
+			trie: &Tag{"A", &Branch{children: Children{
+				0: &Tag{"B", &Empty{}},
+				1: &Tag{"C", &Empty{}},
+				2: &Tag{"D", &Empty{}},
+			}}},
+			path: []string{"A", "C"}, // < this should just be "A"
+		},
+		"branch with wrong account": {
+			trie: &Tag{"A", &Branch{children: Children{
+				0: &Tag{"B", &Empty{}},
+				1: &Tag{"C", &Account{}},
+				2: &Tag{"D", &Empty{}},
+			}}},
+			path: []string{"A", "C"}, // < this should just be "A"
+		},
+		"branch with correct account": {
+			trie: &Tag{"A", &Branch{children: Children{
+				0: &Tag{"B", &Empty{}},
+				1: &Tag{"C", &Account{address: address}},
+				2: &Tag{"D", &Empty{}},
+			}}},
+			path: []string{"A", "C"},
+		},
+		"extension with common prefix": {
+			trie: &Tag{"A", &Extension{
+				path: []Nibble{1, 2, 3},
+				next: &Tag{"B", &Branch{children: Children{
+					3: &Tag{"C", &Empty{}},
+					4: &Tag{"D", &Empty{}},
+				}},
+				}}},
+			path: []string{"A", "B", "C"}, // < this should just be "A" and "B"
+		},
+		"extension without common prefix": {
+			trie: &Tag{"A", &Extension{path: []Nibble{2, 3}}},
+			path: []string{"A"}, // < this should just be nothing
+		},
+		"branch node short path": {
+			trie: &Tag{"A", &Extension{
+				path: addressToNibbles(address), // extension node will exhaust the path
+				next: &Tag{"B", &Branch{}},
+			},
+			},
+			path: []string{"A", "B"},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			ctxt := newNiceNodeContext(t, ctrl)
+
+			root, shared := ctxt.Build(test.trie)
+			accountPresent := false
+			handle := shared.GetViewHandle()
+			if _, err := handle.Get().Visit(ctxt, &root, 0, MakeVisitor(func(n Node, i NodeInfo) VisitResponse {
+				if node, ok := n.(*AccountNode); ok && node.address == address {
+					accountPresent = true
+				}
+				return VisitResponseContinue
+			})); err != nil {
+				t.Fatalf("unexpected error during final visit: %v", err)
+			}
+			handle.Release()
+
+			visitor := NewMockNodeVisitor(ctrl)
+			var last *gomock.Call
+			for i, label := range test.path {
+				// the visitor does not return incorrect accounts,
+				// i.e., include the last node in the path only when
+				// the account does exist.
+				if i < len(test.path)-1 || accountPresent {
```

Comment:
Instead of modifying the test expectations here, the paths defined in the test case structs above should be adjusted.

Reply:
ok, thanks, changed